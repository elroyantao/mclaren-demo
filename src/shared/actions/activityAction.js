import 'isomorphic-fetch'

export function fetchActiviyDefRequest() {
  return {
    type: 'FETCH_ACTIVITYDEF_REQUEST'
  }
}

export function fetchActiviyDefResponse(activities) {
  return {
    type: 'FETCH_ACTIVITYDEF_RESPONSE',
    activities
  }
}

export function fetchActiviyDefFailure() {
  return {
    type: 'FETCH_ACTIVITYDEF_FAILURE'
  }
}

export const fetchActiviyDef = () => (dispatch) => {
  dispatch(fetchActiviyDefRequest())
  return fetch('/api/activity-definition', { method: 'GET' })
    .then((response) => {
      if (!response.ok) throw Error(response.statusText)
      return response.json()
    })
    .then((response) => {
      dispatch(fetchActiviyDefResponse(response))
    })
    .catch(() => {
      dispatch(fetchActiviyDefFailure())
    })
}
