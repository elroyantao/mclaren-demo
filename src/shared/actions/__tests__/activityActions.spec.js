import fetchMock from 'fetch-mock'
import * as actions from '../activityAction'

const fetchAction = actions.fetchActiviyDef()
const dispatch = jest.fn()

describe('activity actions', () => {
  it('fetchActiviyDefRequest()', () => {
    expect(actions.fetchActiviyDefRequest()).toMatchSnapshot()
  })
  it('fetchActiviyDefFailure()', () => {
    expect(actions.fetchActiviyDefFailure()).toMatchSnapshot()
  })
  it('fetchActiviyDefResponse()', () => {
    expect(actions.fetchActiviyDefResponse([1, 2, 3])).toMatchSnapshot()
  })
  describe('fetchActiviyDef() ', () => {
    // fetch.mockResponse(JSON.stringify({ access_token: '12345' }))
    afterEach(() => {
      fetchMock.restore()
      jest.clearAllMocks()
    })
    it('a successful fetch', () => {
      fetchMock.get('/api/activity-definition', {
        data: [1, 2, 3]
      })
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/activity-definition', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
    it('an unsuccessful fetch', () => {
      fetchMock.get('/api/activity-definition', 402)
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/activity-definition', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
  })
})
