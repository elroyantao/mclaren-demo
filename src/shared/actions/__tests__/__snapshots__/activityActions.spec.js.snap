// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity actions fetchActiviyDef()  a successful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_ACTIVITYDEF_REQUEST",
    },
  ],
  Array [
    Object {
      "activities": Object {
        "data": Array [
          1,
          2,
          3,
        ],
      },
      "type": "FETCH_ACTIVITYDEF_RESPONSE",
    },
  ],
]
`;

exports[`activity actions fetchActiviyDef()  an unsuccessful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_ACTIVITYDEF_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "FETCH_ACTIVITYDEF_FAILURE",
    },
  ],
]
`;

exports[`activity actions fetchActiviyDefFailure() 1`] = `
Object {
  "type": "FETCH_ACTIVITYDEF_FAILURE",
}
`;

exports[`activity actions fetchActiviyDefRequest() 1`] = `
Object {
  "type": "FETCH_ACTIVITYDEF_REQUEST",
}
`;

exports[`activity actions fetchActiviyDefResponse() 1`] = `
Object {
  "activities": Array [
    1,
    2,
    3,
  ],
  "type": "FETCH_ACTIVITYDEF_RESPONSE",
}
`;
